lib/galaxy/webapps/galaxy/api/workflows.py

```diff
@@ -65,8 +68,35 @@ def index(self, trans, **kwd):
             item['url'] = url_for( 'workflow', id=encoded_id )
             item['owner'] = wf_sa.stored_workflow.user.username
             rval.append(item)
+        if missing_tools:
+            workflows_missing_tools = []
+            for key, value in enumerate(rval):
+                tool_ids = []
+                workflow_details = self.workflow_contents_manager.workflow_to_dict( trans, self.__get_stored_workflow( trans, value[ 'id' ] ), style='instance' )
+                if 'steps' in workflow_details:
+                    for step in workflow_details[ 'steps' ]:
+                        tool_id = workflow_details[ 'steps' ][ step ][ 'tool_id' ]
+                        if tool_id not in tool_ids and self.__is_missing( trans, tool_id ):
+                            tool_ids.append( tool_id )
+                if len( tool_ids ) > 0:
+                    value[ 'missing_tools' ] = tool_ids
+                    workflows_missing_tools.append( value )
+            return workflows_missing_tools
         return rval
 
+    def __is_missing( self, trans, tool_id ):
+        # First make sure the tool ID A) Is valid, and B) Corresponds to a
+        # toolshed tool.
+        if tool_id is None:
+            return False
+        if 'repos' not in tool_id:
+            return False
+        # It is a valid tool ID, and it is from a toolshed. Check if it's
+        # missing from the toolbox.
+        if trans.app.toolbox.get_tool( tool_id ) is None:
```

Comment:
Actually, I'd move this whole function to the toolbox and give it a signature like ``is_missing_shed_tool(self, tool_id)``